### DISCUSSION TITLE
How do I test that I have configured a method to emit a span?

### DISCUSSION THREAD

**Discussion by jchester** [micronaut-projects/micronaut-tracing#414] [Q&A]
Suppose I have this:

```java
@Singleton
public class HelperExample {
    @NewSpan
    public String create() {
        return "example string";
    }
}
```

I want to be able to test that a span _was_ emitted. Sure, I can launch the app and look at the exported spans, but I want to test-drive behavior, not desk-check it manually.

So I tried:

```java

public class HelperExampleTest {
    @RegisterExtension
    static OpenTelemetryExtension otelTesting = OpenTelemetryExtension.create();

    @Test
    void itEmitsATraceSpan() {
        new HelperExample().create();
        assertThat(otelTesting.getSpans()).isNotEmpty();
    }
}
```

Which doesn't work. I also tried it with the `@Micronaut` annotation, tried it in the context of a full HTTP roundtrip, and tried `@Inject`ing the `InMemorySpanExporter` into the test. No luck with any of these.

Any guidance on how to (preferably unit) test this behavior would be very welcome.

**Comment by luolong** [comment]
Generally speaking, I feel that unit tests are not really the proper place to test such code style / architectural patterns.

I suggest investigating [Trace based testing](https://tracetest.io/) instead as a possible way to assert presence of required spans in a running environment.
